app/connectors/MessageConnector.scala

```diff
@@ -45,58 +47,76 @@ class MessageConnector @Inject() (appConfig: AppConfig, config: Configuration, h
-      "CustomProcessHost"       -> "Digital",
-      HeaderNames.ACCEPT        -> MimeTypes.XML, // can't use ContentTypes.XML because EIS will not accept "application/xml; charset=utf-8"
+      "X-Correlation-Id" -> UUID.randomUUID().toString,
+      "CustomProcessHost" -> "Digital",
+      HeaderNames.ACCEPT -> MimeTypes.XML, // can't use ContentTypes.XML because EIS will not accept "application/xml; charset=utf-8"
       HeaderNames.AUTHORIZATION -> s"Bearer ${details.token}"
     )
 
     implicit val headerCarrier = hc
       .copy(authorization = None, otherHeaders = Seq.empty)
       .withExtraHeaders(requestHeaders: _*)
 
-    def getHeader(header: String): String =
-      headerCarrier
-        .headersForUrl(headerCarrierConfig)(details.url)
-        .find { case (name, _) => name.toLowerCase == header.toLowerCase }
-        .map { case (_, value) => value }
-        .getOrElse("undefined")
-
-    http
-      .POSTString[HttpResponse](details.url, xml.toString)
-      .map { result =>
-        lazy val logMessage =
-          s"""|Posting NCTS message, ${details.routingMessage}
-              |X-Correlation-Id: ${getHeader("X-Correlation-Id")}
-              |${HMRCHeaderNames.xRequestId}: ${getHeader(HMRCHeaderNames.xRequestId)}
-              |X-Message-Type: ${getHeader("X-Message-Type")}
-              |X-Message-Sender: ${getHeader("X-Message-Sender")}
-              |Accept: ${getHeader("Accept")}
-              |CustomProcessHost: ${getHeader("CustomProcessHost")}
-              |Response status: ${result.status}
+    try {
+      http
+        .POSTString[HttpResponse](details.url, xml.toString)
+        .map { result =>
+          lazy val logMessage =
+            s"""|Posting NCTS message, ${details.routingMessage}
+                |X-Correlation-Id: ${getHeader("X-Correlation-Id", details.url)}
+                |${HMRCHeaderNames.xRequestId}: ${getHeader(HMRCHeaderNames.xRequestId,  details.url)}
+                |X-Message-Type: ${getHeader("X-Message-Type",  details.url)}
+                |X-Message-Sender: ${getHeader("X-Message-Sender",  details.url)}
+                |Accept: ${getHeader("Accept",  details.url)}
+                |CustomProcessHost: ${getHeader("CustomProcessHost",  details.url)}
+                |Response status: ${result.status}
               """.stripMargin
 
-        if (Status.isServerError(result.status) || result.status == Status.FORBIDDEN)
-          logger.warn(logMessage)
-        else
-          logger.info(logMessage)
+          if (Status.isServerError(result.status) || result.status == Status.FORBIDDEN)
+            logger.warn(logMessage)
+          else
+            logger.info(logMessage)
 
-        result
-      }
-      .recover {
-        case e: Exception => {
-          val message = s"${details.url} failed to retrieve data with message ${e.getMessage}"
-          logger.warn(message)
-          HttpResponse(Status.INTERNAL_SERVER_ERROR, message)
+          result
```

Comment:
I think converting a `recover` block on a `Future` into a `try..catch` will swallow exceptions

Reply:
I had to change this because it didn't appear to be recovering when exception were thrown correctly in the tests. It'll catch any subtype of Exception.

Reply:
That's because the test is triggering an exception to be thrown using `thenThrow`, which I don't think will happen in practice as the HTTP request should happen on another thread (unless http-verbs does something very odd).

I think you can use `thenReturn(Future.failed(new RuntimeException))` in your test to simulate that instead.

This has drawn attention to one thing about the existing code that should probably be changed though - it catches `Throwable` which could stop a VM which has triggered an unrecoverable error from exiting properly.

It should probably do something like this instead:
```scala
.recover {
  case NonFatal(e) =>
   // etc.
}
```

Reply:
Looking at the code that was there before, when compiling it there was the following warning which led me down the garden path it seems 😃 

```
[warn] /home/dthornton/hmrc/transit-movements-trader-eis-router/it/connectors/MessageConnectorSpec.scala:177:33: dead code following this construct
[warn]         when(http.POSTString(any(), any(), any())(any(), any(), any())).thenReturn(failed(new RuntimeException("Simulated timeout")))
```

I see what you mean about returning the Future.failed (had a look at HttpVerb). I'll update this tomorrow.

`  def mapErrors(
    httpMethod: String,
    url       : String,
    f         : Future[HttpResponse]
  )(implicit
    ec: ExecutionContext
  ): Future[HttpResponse] =
    f.recoverWith {
      case e: TimeoutException => Future.failed(new GatewayTimeoutException(gatewayTimeoutMessage(httpMethod, url, e)))
      case e: ConnectException => Future.failed(new BadGatewayException(badGatewayMessage(httpMethod, url, e)))
    }`